File: anonymous-data/offchain/evolutionsdk/anonymous-data.ts
import {
  Lucid,
  Koios,
  applyParamsToScript,
  Constr,
  Data,
  fromHex,
  toUnit,
  validatorToAddress,
  validatorToScriptHash,
  type LucidEvolution,
  type Validator,
} from "@evolution-sdk/lucid";
import blake2b from "blake2b";
import blueprint from "../../onchain/aiken/plutus.json" with { type: "json" };

// ----------------------------------------------------------------------------
// Anonymous data commit/reveal — Evolution SDK port.
//
//   ID = blake2b_256(pkh || nonce)
//
// Commit (mint): mints exactly one token, asset_name = ID, sent to the script
//   address with an inline datum (the user's opaque payload).
// Reveal (spend): consumes the committed UTxO with redeemer = nonce; the spender
//   is required-signed and their pkh + nonce must reproduce ID.
// ----------------------------------------------------------------------------

const KOIOS_URL = "https://preprod.koios.rest/api/v1";

function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}

function hexToBytes(hex: string): Uint8Array {
  return new Uint8Array(hex.match(/.{1,2}/g)!.map((b) => parseInt(b, 16)));
}

// blake2b_256(pkh || nonce)
function computeIdHex(pkhHex: string, nonceHex: string): string {
  const pkh = hexToBytes(pkhHex);
  const nonce = hexToBytes(nonceHex);
  const combined = new Uint8Array(pkh.length + nonce.length);
  combined.set(pkh);
  combined.set(nonce, pkh.length);
  const hash = blake2b(blake2b.BYTES).update(combined).digest();
  return bytesToHex(hash);
}

async function setup() {
  const lucid = await Lucid(new Koios(KOIOS_URL), "Preprod");
  // Wallet selection is up to the caller — `commit` / `reveal` accept a wallet path.
  return { lucid };
}

function loadValidator(): { validator: Validator; policyId: string; scriptAddress: string } {
  const compiledCode = blueprint.validators[0].compiledCode;
  // No validator parameters.
  const script = applyParamsToScript(compiledCode, []);
  const validator: Validator = { type: "PlutusV3", script };
  const policyId = validatorToScriptHash(validator);
  const scriptAddress = validatorToAddress("Preprod", validator);
  return { validator, policyId, scriptAddress };
}

async function commit(walletFile: string, nonceHex: string, dataHex: string) {
  const { lucid } = await setup();
  const mnemonic = Deno.readTextFileSync(walletFile).trim();
  lucid.selectWallet.fromSeed(mnemonic);

  const address = await lucid.wallet().address();
  const { paymentCredential } = (await import("@evolution-sdk/lucid")).getAddressDetails(address);
  if (!paymentCredential) throw new Error("No payment credential on wallet");

  const idHex = computeIdHex(paymentCredential.hash, nonceHex);
  console.log(`ID = blake2b_256(pkh || nonce) = ${idHex}`);

  const { validator, policyId, scriptAddress } = loadValidator();
  const unit = toUnit(policyId, idHex);

  // Mint redeemer for the policy is the id (ByteArray).
  const mintRedeemer = Data.to(idHex);
  // Inline datum on the script output is the user's data (opaque ByteArray).
  const datum = Data.to(dataHex);

  const tx = await lucid
    .newTx()
    .mintAssets({ [unit]: 1n }, mintRedeemer)
    .attach.MintingPolicy(validator)
    .pay.ToContract(scriptAddress, { kind: "inline", value: datum }, { [unit]: 1n })
    .complete();

  const signed = await tx.sign.withWallet().complete();
  const txHash = await signed.submit();
  console.log(`Commit submitted. Tx: ${txHash}`);
  console.log(`Reveal later with:  reveal ${walletFile} ${nonceHex}`);
}

async function reveal(walletFile: string, nonceHex: string) {
  const { lucid } = await setup();
  const mnemonic = Deno.readTextFileSync(walletFile).trim();
  lucid.selectWallet.fromSeed(mnemonic);

  const address = await lucid.wallet().address();
  const { paymentCredential } = (await import("@evolution-sdk/lucid")).getAddressDetails(address);
  if (!paymentCredential) throw new Error("No payment credential on wallet");

  const idHex = computeIdHex(paymentCredential.hash, nonceHex);
  const { validator, policyId, scriptAddress } = loadValidator();
  const unit = toUnit(policyId, idHex);

  const utxos = await lucid.utxosAt(scriptAddress);
  const utxo = utxos.find((u) => (u.assets[unit] ?? 0n) === 1n);
  if (!utxo) throw new Error(`Committed UTxO with unit ${unit} not found`);

  const spendRedeemer = Data.to(nonceHex);

  // Note: we do NOT burn the token. The mint handler enforces
  // `token_minted(..., +1)` and would reject a -1 burn. The spend handler
  // imposes no constraint on where the token goes, so we send the full
  // UTxO value (token + lovelace) back to the spender's wallet via change.
  const tx = await lucid
    .newTx()
    .collectFrom([utxo], spendRedeemer)
    .attach.SpendingValidator(validator)
    .addSigner(address)
    .complete();

  const signed = await tx.sign.withWallet().complete();
  const txHash = await signed.submit();
  console.log(`Reveal submitted. Tx: ${txHash}`);
}

if (import.meta.main) {
  const [cmd, ...args] = Deno.args;
  if (!cmd) {
    console.log(
      "Usage:\n" +
        "  commit <wallet.json> <nonce_hex> <data_hex>\n" +
        "  reveal <wallet.json> <nonce_hex>\n",
    );
  } else if (cmd === "commit") {
    if (args.length !== 3) console.error("Usage: commit <wallet.json> <nonce_hex> <data_hex>");
    else await commit(args[0], args[1], args[2]);
  } else if (cmd === "reveal") {
    if (args.length !== 2) console.error("Usage: reveal <wallet.json> <nonce_hex>");
    else await reveal(args[0], args[1]);
  } else {
    console.log("Unknown command");
  }
}

File: anonymous-data/offchain/ccl-java/AnonymousData.java
/// usr/bin/env jbang "$0" "$@" ; exit $?
///
// @formatter:off
//JAVA 24+

//DEPS com.bloxbean.cardano:cardano-client-lib:0.8.0-pre4
//DEPS com.bloxbean.cardano:cardano-client-backend-blockfrost:0.8.0-pre4
//DEPS com.bloxbean.cardano:aiken-java-binding:0.1.0
// @formatter:on

import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

import com.bloxbean.cardano.client.account.Account;
import com.bloxbean.cardano.client.address.Address;
import com.bloxbean.cardano.client.address.AddressProvider;
import com.bloxbean.cardano.client.api.UtxoSupplier;
import com.bloxbean.cardano.client.api.exception.ApiException;
import com.bloxbean.cardano.client.api.model.Amount;
import com.bloxbean.cardano.client.api.model.Utxo;
import com.bloxbean.cardano.client.backend.api.BackendService;
import com.bloxbean.cardano.client.backend.api.DefaultUtxoSupplier;
import com.bloxbean.cardano.client.backend.blockfrost.service.BFBackendService;
import com.bloxbean.cardano.client.common.model.Network;
import com.bloxbean.cardano.client.common.model.Networks;
import com.bloxbean.cardano.client.crypto.Blake2bUtil;
import com.bloxbean.cardano.client.function.helper.SignerProviders;
import com.bloxbean.cardano.client.plutus.blueprint.PlutusBlueprintLoader;
import com.bloxbean.cardano.client.plutus.blueprint.PlutusBlueprintUtil;
import com.bloxbean.cardano.client.plutus.blueprint.model.PlutusContractBlueprint;
import com.bloxbean.cardano.client.plutus.blueprint.model.PlutusVersion;
import com.bloxbean.cardano.client.plutus.spec.BytesPlutusData;
import com.bloxbean.cardano.client.plutus.spec.PlutusData;
import com.bloxbean.cardano.client.plutus.spec.PlutusScript;
import com.bloxbean.cardano.client.quicktx.QuickTxBuilder;
import com.bloxbean.cardano.client.quicktx.ScriptTx;
import com.bloxbean.cardano.client.quicktx.TxResult;
import com.bloxbean.cardano.client.transaction.spec.Asset;
import com.bloxbean.cardano.client.util.HexUtil;

/**
 * Anonymous data commit/reveal use-case.
 *
 *   ID = blake2b_256(pkh || nonce)
 *
 * Commit (mint): mint exactly one token with asset_name = ID, send it to the
 *   script address with an inline datum holding arbitrary user data.
 * Reveal (spend): spend that UTxO supplying the nonce as redeemer; the spender
 *   must be a signer whose pkh can reproduce the committed ID.
 *
 * The same script provides both the mint policy and the spend validator.
 */
public class AnonymousData {

        static BackendService backendService =
                        new BFBackendService("http://localhost:8080/api/v1/", "Dummy Key");
        static UtxoSupplier utxoSupplier = new DefaultUtxoSupplier(backendService.getUtxoService());

        static String mnemonic = "test test test test test test test test test test test test test test test test test test test test test test test sauce";
        static Network network = Networks.testnet();
        static Account owner = Account.createFromMnemonic(network, mnemonic);
        static Address ownerAddress = owner.getBaseAddress();
        static QuickTxBuilder quickTxBuilder = new QuickTxBuilder(backendService);

        // The validator has no parameters; the same compiled script serves both
        // the mint policy and the spending validator. Its hash is the policy id.
        static PlutusScript plutusScript = loadPlutusScript();
        static String policyId = computePolicyId(plutusScript);
        static Address scriptAddress = AddressProvider.getEntAddress(plutusScript, network);

        private static String computePolicyId(PlutusScript script) {
                try {
                        return script.getPolicyId();
                } catch (Exception e) {
                        throw new RuntimeException("Failed to compute policyId", e);
                }
        }

        // Test inputs.
        static String NONCE_HEX = "deadbeef";
        static String DATA_HEX = HexUtil.encodeHexString("anonymous-payload".getBytes(StandardCharsets.UTF_8));

        public static void main(String[] args) throws ApiException, InterruptedException {
                byte[] pkh = ownerAddress.getPaymentCredentialHash().get();
                byte[] nonce = HexUtil.decodeHexString(NONCE_HEX);
                byte[] id = blake2b256(concat(pkh, nonce));
                String idHex = HexUtil.encodeHexString(id);
                System.out.println("ID = blake2b_256(pkh || nonce) = " + idHex);

                TxResult commitTx = commit(id, HexUtil.decodeHexString(DATA_HEX));
                System.out.println("COMMIT result: successful=" + commitTx.isSuccessful()
                                + " txHash=" + commitTx.getTxHash());
                if (!commitTx.isSuccessful())
                        throw new AssertionError("Commit failed");
                waitForUtxoWithToken(idHex, 60);

                TxResult revealTx = reveal(id, nonce);
                System.out.println("REVEAL result: successful=" + revealTx.isSuccessful()
                                + " txHash=" + revealTx.getTxHash());
                if (!revealTx.isSuccessful())
                        throw new AssertionError("Reveal failed");
        }

        private static TxResult commit(byte[] id, byte[] data) throws ApiException {
                String idHex = HexUtil.encodeHexString(id);

                // Mint redeemer for the policy is the id (ByteArray).
                PlutusData mintRedeemer = BytesPlutusData.of(id);
                // Inline datum on the script output is the user's data (opaque).
                PlutusData inlineDatum = BytesPlutusData.of(data);

                Asset asset = Asset.builder()
                                .name("0x" + idHex)
                                .value(BigInteger.ONE)
                                .build();

                // Two-step: declare the policy script (so we can use the policyId-string
                // overload of mintAsset that supports an inline datum on the output).
                ScriptTx scriptTx = new ScriptTx()
                                .mintAsset(policyId, List.of(asset), mintRedeemer,
                                                scriptAddress.getAddress(), inlineDatum)
                                .attachMintValidator(plutusScript);

                return quickTxBuilder.compose(scriptTx)
                                .feePayer(ownerAddress.getAddress())
                                .withSigner(SignerProviders.signerFrom(owner))
                                .completeAndWait();
        }

        private static TxResult reveal(byte[] id, byte[] nonce) throws ApiException {
                String idHex = HexUtil.encodeHexString(id);
                Utxo target = utxoSupplier.getAll(scriptAddress.getAddress()).stream()
                                .filter(u -> u.getAmount().stream().anyMatch(a ->
                                                (policyId + idHex).equalsIgnoreCase(a.getUnit())))
                                .findFirst()
                                .orElseThrow(() -> new IllegalStateException("Committed UTxO not found"));

                long slot = backendService.getBlockService().getLatestBlock().getValue().getSlot();

                // Spend redeemer is the nonce ByteArray.
                PlutusData spendRedeemer = BytesPlutusData.of(nonce);

                // Note: we do NOT burn the token. The contract's mint handler enforces
                // `token_minted(..., +1)` and would reject a -1 burn. The spend handler
                // imposes no constraint on where the token goes, so we forward the
                // entire UTxO value (token + lovelace) to the owner via change.
                ScriptTx scriptTx = new ScriptTx()
                                .collectFrom(List.of(target), spendRedeemer)
                                .attachSpendingValidator(plutusScript)
                                .withChangeAddress(ownerAddress.getAddress());

                return quickTxBuilder.compose(scriptTx)
                                .validFrom(slot - 5)
                                .validTo(slot + 5)
                                .feePayer(ownerAddress.getAddress())
                                .withSigner(SignerProviders.signerFrom(owner))
                                .withRequiredSigners(ownerAddress)
                                .completeAndWait();
        }

        private static void waitForUtxoWithToken(String idHex, int timeoutSec) throws InterruptedException {
                String unit = policyId + idHex.toLowerCase();
                for (int i = 0; i < timeoutSec; i++) {
                        boolean found = utxoSupplier.getAll(scriptAddress.getAddress()).stream()
                                        .flatMap(u -> u.getAmount().stream())
                                        .anyMatch(a -> unit.equalsIgnoreCase(a.getUnit()));
                        if (found) {
                                System.out.println("Committed UTxO indexed after " + i + "s");
                                return;
                        }
                        Thread.sleep(1000);
                }
                System.out.println("Timed out waiting for committed UTxO after " + timeoutSec + "s");
        }

        private static PlutusScript loadPlutusScript() {
                Path plutusJson = Paths.get(System.getProperty("user.dir"),
                                "..", "..", "onchain", "aiken", "plutus.json");
                PlutusContractBlueprint blueprint = PlutusBlueprintLoader.loadBlueprint(plutusJson.toFile());
                String compiledCode = blueprint.getValidators().getFirst().getCompiledCode();
                return PlutusBlueprintUtil.getPlutusScriptFromCompiledCode(compiledCode, PlutusVersion.v3);
        }

        private static byte[] concat(byte[] a, byte[] b) {
                byte[] r = new byte[a.length + b.length];
                System.arraycopy(a, 0, r, 0, a.length);
                System.arraycopy(b, 0, r, a.length, b.length);
                return r;
        }

        private static byte[] blake2b256(byte[] in) {
                return Blake2bUtil.blake2bHash256(in);
        }
}
